### DISCUSSION TITLE
Spatial initial condition from file

### DISCUSSION THREAD

**Discussion by Khaksarkn** [idaholab/moose#29595] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi,
I was wondering if there is a way of specifying the initial distribution of a variable based on grid data. Say the data is available on a grid from a csv file like this (as a function of space; but it's not an expression function, it more like data collected from measurement):
```
x, y, z, temperature
0.0, 0.0, 0.0, 5.0
10.0, 0.0, 0.0, 4.0
10.0, 10.0, 0.0, 3.0
0.0, 10.0, 0.0, 6.0
```
and if it can interpolate on the nodes like a ```scattered interpolant```?
I searched the functions for a function that reads from a file to use a ```Function IC``` (something like a ```ParsedFileFunction```). but I didn't find one. 
Or if there is a way to specify on the mesh (in the mesh module)

I would appreciate it if you could guide me with this.

This is a sample input file I was trying to modify for this:

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 5
  ny = 5
  xmin = 0
  xmax = 10
  ymin = 0
  ymax = 10
[]

[Variables]
  [temperature]
    order = FIRST
    family = LAGRANGE
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = temperature
  []
  [timederiv]
    type = TimeDerivative
    variable = temperature
  []
[]

[BCs]
  [bottomofthesample]
    type = DirichletBC
    variable = temperature
    boundary = bottom
    value = 1
  []
  
  [topofthesample]
    type = DirichletBC
    variable = temperature
    boundary = top
    value = 0
  []
[]

[Executioner]
  type = Transient
  num_steps = 100
  dt = 1
  solve_type = 'NEWTON'
[]

[Outputs]
  execute_on = 'TIMESTEP_BEGIN'
  exodus = true
[]


```

**Comment by GiudGiud** [comment, marked as answer]
Hello

If the grid is regular the Piecewise(Multi)Constant Piecewise(Multi)Linear functions can read it, and the FunctionIC can be used:
https://mooseframework.inl.gov/source/functions/PiecewiseMultiConstant.html
https://mooseframework.inl.gov/source/functions/PiecewiseMultiLinear.html

then this
https://mooseframework.inl.gov/source/ics/FunctionIC.html

If not, use a PropertyReadFile user object to read the file (using the "voronoi" mode), then a PiecewiseConstantFromCSV function and a FunctionIC
https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html
https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html
